File: content/LLD/Reference-Code/14. Builder Design Pattern/good_example.py
class Laptop:
    processor = None
    ram = None
    graphic_card = None
    color = None
    screen_size = None

    def display_specs(self):
        if self.processor:
            print(f"processor = {self.processor}")
        if self.ram:
            print(f"Ram = {self.ram}")
        if self.graphic_card:
            print(f"Graphic Card = {self.graphic_card}")
        if self.color:
            print(f"Color = {self.color}")
        if self.screen_size:
            print(f"Screen Size = {self.screen_size}")


class LaptopBuilder:
    def __init__(self):
        self.__laptop = Laptop()

    def set_processor(self, processor: str):
        self.__laptop.processor = processor
        return self

    def set_ram(self, ram: str):
        self.__laptop.ram = ram
        return self

    def set_graphic_card(self, graphic_card: str):
        self.__laptop.graphic_card = graphic_card
        return self

    def set_color(self, color: str):
        self.__laptop.color = color
        return self

    def set_screen_size(self, screen_size: str):
        self.__laptop.screen_size = screen_size
        return self

    def build(self):
        return self.__laptop


l = LaptopBuilder().set_processor("i5-43423").set_ram("5").set_color("black").build()
l.display_specs()

File: content/LLD/Reference-Code/14. Builder Design Pattern/GoodExample.java
public class GoodExample {
    static class Laptop {
        private String processor;
        private String ram;
        private String graphicCard;
        private String color;
        private String screenSize;

        public void displaySpecs() {
            if (this.processor != null) {
                System.out.println("processor = " + this.processor);
            }
            if (this.ram != null) {
                System.out.println("Ram = " + this.ram);
            }
            if (this.graphicCard != null) {
                System.out.println("Graphic Card = " + this.graphicCard);
            }
            if (this.color != null) {
                System.out.println("Color = " + this.color);
            }
            if (this.screenSize != null) {
                System.out.println("Screen Size = " + this.screenSize);
            }
        }
    }

    static class LaptopBuilder {
        private final Laptop laptop = new Laptop();

        public LaptopBuilder setProcessor(String processor) {
            this.laptop.processor = processor;
            return this;
        }

        public LaptopBuilder setRam(String ram) {
            this.laptop.ram = ram;
            return this;
        }

        public LaptopBuilder setGraphicCard(String graphicCard) {
            this.laptop.graphicCard = graphicCard;
            return this;
        }

        public LaptopBuilder setColor(String color) {
            this.laptop.color = color;
            return this;
        }

        public LaptopBuilder setScreenSize(String screenSize) {
            this.laptop.screenSize = screenSize;
            return this;
        }

        public Laptop build() {
            return this.laptop;
        }
    }

    public static void main(String[] args) {
        Laptop l = new LaptopBuilder()
            .setProcessor("i5-43423")
            .setRam("5")
            .setColor("black")
            .build();
        l.displaySpecs();
    }
}
